shoes-core/lib/shoes/download.rb

```diff
@@ -100,26 +103,25 @@ def default_error_proc
       end
     end
 
-    def content_length_proc
-      lambda do |content_length|
-        download_started(content_length)
-        eval_block(@progress_blk, self)
+    def try_progress(size)
+      if should_mark_progress?(size)
+        @transferred = size
+        mark_progress
       end
     end
 
-    def progress_proc
-      lambda do |size|
-        if !content_length.nil? &&
-           (size - transferred) > (content_length / UPDATE_STEPS) &&
-           !@gui.busy?
-          @gui.busy = true
-          eval_block(@progress_blk, self)
-          @transferred = size
-        end
-      end
+    def mark_progress
+      @gui.busy = true
+      eval_block(@progress_blk, self)
     end
 
-    def finish_download(_download_data)
+    def should_mark_progress?(size)
+      !content_length.nil? &&
+        (size - transferred) > (content_length / UPDATE_STEPS) &&
+        !@gui.busy?
```

Comment:
while you're in here - maybe you remeber/know I just have the vague feeling that something went wrong whenever we updated progress too fast and then stuff crashed... if you remember, maybe a little why style comment for the real reason/link to issue/whatever?